lib/segment/src/problems/unindexed_field.rs

```diff
@@ -0,0 +1,330 @@
+
+        match solutions.len() {
+            0 => unreachable!(
+                "Cannot create a solution without a field schema, protected by try_new()"
+            ),
+            1 => Solution::Immediate(solutions.pop().unwrap()),
+            _ => Solution::ImmediateChoice(solutions),
+        }
+    }
+}
+
+fn infer_schema_from_match_value(value: &MatchValue) -> PayloadFieldSchema {
+    match &value.value {
+        crate::types::ValueVariants::Keyword(_string) => {
+            PayloadFieldSchema::FieldType(PayloadSchemaType::Keyword)
+        }
+        crate::types::ValueVariants::Integer(_integer) => {
+            PayloadFieldSchema::FieldType(PayloadSchemaType::Integer)
+        }
+        crate::types::ValueVariants::Bool(_boolean) => {
+            PayloadFieldSchema::FieldType(PayloadSchemaType::Bool)
+        }
+    }
+}
+
+fn infer_schema_from_any_variants(value: &AnyVariants) -> PayloadFieldSchema {
+    match value {
+        AnyVariants::Keywords(_strings) => {
+            PayloadFieldSchema::FieldType(PayloadSchemaType::Keyword)
+        }
+        AnyVariants::Integers(_integers) => {
+            PayloadFieldSchema::FieldType(PayloadSchemaType::Integer)
+        }
+    }
+}
+
+fn infer_schema_from_field_condition(field_condition: &FieldCondition) -> Vec<PayloadFieldSchema> {
+    match field_condition {
+        FieldCondition {
+            r#match: Some(r#match),
+            ..
+        } => vec![match r#match {
+            Match::Value(match_value) => infer_schema_from_match_value(match_value),
+            Match::Text(_match_text) => {
+                PayloadFieldSchema::FieldParams(PayloadSchemaParams::Text(TextIndexParams {
+                    r#type: TextIndexType::Text,
+                    tokenizer: TokenizerType::default(),
+                    min_token_len: None,
+                    max_token_len: None,
+                    lowercase: None,
+                }))
+            }
+            Match::Any(match_any) => infer_schema_from_any_variants(&match_any.any),
+            Match::Except(match_except) => infer_schema_from_any_variants(&match_except.except),
+        }],
+        FieldCondition {
+            range: Some(_range),
+            ..
+        } => vec![
+            PayloadFieldSchema::FieldType(PayloadSchemaType::Integer),
+            PayloadFieldSchema::FieldType(PayloadSchemaType::Float),
+        ],
+        FieldCondition {
+            geo_bounding_box: Some(_),
+            ..
+        }
+        | FieldCondition {
+            geo_radius: Some(_),
+            ..
+        }
+        | FieldCondition {
+            geo_polygon: Some(_),
+            ..
+        } => vec![PayloadFieldSchema::FieldType(PayloadSchemaType::Geo)],
+        FieldCondition {
+            key: _,
+            r#match: None,
+            range: None,
+            geo_bounding_box: None,
+            geo_radius: None,
+            geo_polygon: None,
+            values_count: _,
+        } => vec![],
+    }
```

Comment:
:pick: Nitpick: could be rewritten in this style. This way, `cargo check` would complain about unused variables if you miss handling of some field.

```rust
let FieldCondition {
    r#match,
    range,
    geo_bounding_box,
    …
} = field_condition;

if let Some(r#match) = r#match {
    vec![…]
} else if let Some(range) = range {
    vec![…]
} … {
    vec![…]
} else {
    vec![]
}
```

Reply:
rewritten :)